implementation/src/main/java/io/smallrye/config/TracedPropertiesConfigSource.java

```diff
@@ -0,0 +1,47 @@
+package io.smallrye.config;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.LineNumberReader;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import io.smallrye.config.common.MapBackedConfigSource;
+import io.smallrye.config.common.utils.ConfigSourceUtil;
+
+public class TracedPropertiesConfigSource extends MapBackedConfigSource implements TracedConfigSource {
+    private static final long serialVersionUID = 3969193230016229478L;
+
+    private static final String NAME_PREFIX = "TracedPropertiesConfigSource[source=";
+
+    private final URL url;
+    private final Map<String, Integer> tracedProperties = new HashMap<>();
+
+    public TracedPropertiesConfigSource(URL url) throws IOException {
+        super(NAME_PREFIX + url.toString() + "]", ConfigSourceUtil.urlToMap(url));
+        this.url = url;
+        generateTracingInfo(url);
+    }
+
+    private void generateTracingInfo(URL locationOfProperties) throws IOException {
+        try (LineNumberReader reader = new LineNumberReader(
+                new InputStreamReader(locationOfProperties.openStream(), StandardCharsets.UTF_8))) {
+
+            String line;
+            while ((line = reader.readLine()) != null) {
+                findName(line).ifPresent(name -> tracedProperties.put(name, reader.getLineNumber()));
+            }
+        }
+    }
+
+    private Optional<String> findName(String line) {
+        return getProperties().keySet().stream().filter(line::contains).findFirst();
```

Comment:
This linear search introduces a polynomial cost to property loading.  10 keys takes 100 iterations to search; 1000 keys takes 1,000,000 iterations.

Reply:
See previous comments.